plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts

```diff
@@ -229,51 +300,105 @@ export class DatabaseNotificationsStore implements NotificationsStore {
     id: string;
     notification: Notification;
   }) {
-    const query = this.db('notification')
-      .where('id', options.id)
-      .where('user', options.notification.user);
-
-    await query.update({
+    const updateColumns = {
       title: options.notification.payload.title,
       description: options.notification.payload.description,
       link: options.notification.payload.link,
       topic: options.notification.payload.topic,
       updated: new Date(),
       severity: options.notification.payload.severity,
       read: null,
-    });
+    };
+
+    const notificationQuery = this.db('notification')
+      .where('id', options.id)
+      .where('user', options.notification.user);
+    const broadcastQuery = this.db('broadcast').where('id', options.id);
+
+    await Promise.all([
+      notificationQuery.update(updateColumns),
+      broadcastQuery.update({ ...updateColumns, read: undefined }),
+    ]);
 
     return await this.getNotification(options);
   }
 
   async getNotification(options: { id: string }): Promise<Notification | null> {
-    const rows = await this.db('notification')
+    const rows = await this.db
+      .select('*')
+      .from(
+        this.db('notification')
+          .select(NOTIFICATION_COLUMNS)
+          .unionAll([this.getBroadcastUnion()])
+          .as('notifications'),
+      )
       .where('id', options.id)
-      .select()
       .limit(1);
     if (!rows || rows.length === 0) {
       return null;
     }
     return this.mapToNotifications(rows)[0];
   }
 
+  private markReadSaved = async (
+    ids: string[],
+    user: string,
+    read?: Date | null,
+    saved?: Date | null,
+  ) => {
+    await this.db('notification')
+      .whereIn('id', ids)
+      .where('user', user)
+      .update({ read, saved });
+
+    const broadcasts = this.mapToNotifications(
+      await this.db('broadcast').whereIn('id', ids).select(),
+    );
+
+    if (broadcasts.length > 0)
+      if (!this.isSQLite) {
+        await this.db('broadcast_user_status')
+          .insert(
+            broadcasts.map(b => ({
+              broadcast_id: b.id,
+              user,
+              read,
+              saved,
+            })),
+          )
+          .onConflict(['broadcast_id', 'user'])
```

Comment:
Is there a need to have two completely different flows per engine?

The 2nd one for the SQLite is not so nice but works and leaves us with less code to maintain.

Reply:
I think sqlite is the exception here as this doesn't support this. This way is much faster and the proper way to do this.